_11ty/apply-csp.js

```diff
@@ -63,20 +65,35 @@ const addCspHash = async (rawContent, outputPath) => {
       hashes.push.apply(hashes, AUTO_RELOAD_SCRIPTS);
     }
 
-    const csp = dom.window.document.querySelector(
-      "meta[http-equiv='Content-Security-Policy']"
-    );
-    if (!csp) {
-      return content;
-    }
-    csp.setAttribute(
-      "content",
-      csp.getAttribute("content").replace("HASHES", hashes.join(" "))
-    );
-
     content = dom.serialize();
+
+    // write CSP Policy in headers file
+    const headersPath = "./_site/_headers";
+    const filePath = outputPath.replace("_site/", "/"); // _site/blog/index.html ->  /blog/index.html
+    const filePathPrettyURL = filePath.slice(0, -10); // blog/index.html ->  /blog/
+    try {
+      const headers = fs.readFileSync(headersPath, { encoding: "utf-8" });
+      const regExp = /(# \[custom headers\]\n)([\s\S]*)(# \[end custom headers\])/;
+      const match = headers.match(regExp);
+      if (!match) {
+        throw `Check your _headers file. I couldn't find the text block for the custom headers:
+          # [custom headers]
+          # this text will be replaced by apply-csp.js plugin
+          # [end custom headers]`;
+      }
+      const oldCustomHeaders = headers.match(regExp)[2].toString();
+      const CSPPolicy = `Content-Security-Policy: ${CSP.apply().regular.replace("HASHES", hashes.join(" "))}`;
+      // write headers for full path (/blog/index.html) and pretty url (/blog/)
+      const newCustomHeaders = oldCustomHeaders.concat(
+        "\n", filePath, "\n  ", CSPPolicy,
+        "\n", filePathPrettyURL, "\n  ", CSPPolicy);
+      fs.writeFileSync(headersPath, headers.replace(regExp, `$1${newCustomHeaders}\n$3`));
```

Comment:
I'm not a fan that this relies on sync writes, but it is fine :)

Reply:
if I'm not sure if an async call can get some headers lost along the way.

Reply:
I changed the CSP policy implementation for 404 pages. Initially it didn't work for generic pages not found, maybe caching problems, so I thought it needed to be managed with a meta tag. But further tests gave positive results and so I used the headers for 404 pages as well.